File: app/src/main/java/com/yeraygarcia/recipes/database/dao/RecipeDao.java
package com.yeraygarcia.recipes.database.dao;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Query;

import com.yeraygarcia.recipes.database.entity.Recipe;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Dao
public abstract class RecipeDao extends BaseDao<Recipe> {

    @Query("DELETE FROM recipe")
    public abstract void deleteAll();

    @Query("SELECT * FROM recipe ORDER BY name ASC")
    public abstract LiveData<List<Recipe>> findAll();

    @Query("SELECT count(1) = 0 AS is_empty FROM recipe")
    public abstract boolean isEmpty();

    @Query("SELECT * FROM recipe WHERE id IN (SELECT DISTINCT recipe_id FROM shopping_list_item)")
    public abstract LiveData<List<Recipe>> findRecipesInShoppingList();

    @Query("DELETE FROM recipe WHERE id NOT IN (:ids)")
    abstract void deleteIfIdNotIn(List<UUID> ids);

    public void deleteIfNotIn(List<Recipe> entities) {
        List<UUID> ids = new ArrayList<>();
        for (Recipe entity : entities) {
            ids.add(entity.getId());
        }
        deleteIfIdNotIn(ids);
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/database/dao/RecipeDao.kt
package com.yeraygarcia.recipes.database.dao

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Query
import com.yeraygarcia.recipes.database.entity.Recipe
import java.util.*

@Dao
abstract class RecipeDao : BaseDao<Recipe>() {

    @get:Query("SELECT count(1) = 0 AS is_empty FROM recipe")
    abstract val isEmpty: Boolean

    @Query("DELETE FROM recipe")
    abstract fun deleteAll()

    @Query("SELECT * FROM recipe ORDER BY name ASC")
    abstract fun findAll(): LiveData<List<Recipe>>

    @Query("SELECT * FROM recipe WHERE id = :id")
    abstract fun findById(id: UUID): LiveData<Recipe>

    @Query("SELECT * FROM recipe WHERE id = :id")
    abstract fun findRawById(id: UUID): Recipe

    @Query("SELECT * FROM recipe WHERE id IN (SELECT DISTINCT recipe_id FROM shopping_list_item)")
    abstract fun findRecipesInShoppingList(): LiveData<List<Recipe>>

    @Query("DELETE FROM recipe WHERE id NOT IN (:ids)")
    internal abstract fun deleteIfIdNotIn(ids: List<UUID>)

    fun deleteIfNotIn(entities: List<Recipe>) {
        val ids = ArrayList<UUID>()
        for (entity in entities) {
            ids.add(entity.id)
        }
        deleteIfIdNotIn(ids)
    }
}
